### PR TITLE
Support for configurable debug origin/port (issue 1429)

### PR DESCRIPTION
**Pull request by bparadie**
This PR adds support for configurable debug origin/port information as discussed in https://github.com/facebook/react-native/issues/1429 .

#### Summary 

Here is a brief summary of my changes:
- I added a new target called `ReactWithoutPackager` to `React.xcodeproj` that is a copy of the `React` target but does not execute the `launchPackager.command` script.
- `UIExplorer` now links against `libReactWithoutPackager.a`, which attracts the `ReactWithoutPackager` target w/o `launchPackager.command`.
- I added support for `--origin` parameter to `packager.sh`
- I added a run script to `UIExplorer` that launches `launchPackager.command` with `--origin` and `--port` parameters.
- Because I am now passing `--origin` and `--port` as parameters to `launchPackager.command` and `open` w/t `--args` does not seem to work with `Terminal`, I changed that code to using `osascript` instead.

#### Changing the port number 

This is a screenshot after changing the port number to `8086` in `Info.plist` and rebuilding `UIExplorer`:

screen shot 2015-06-06 at 4 46 09 pm

#### Backwards compatibility

I decided to make this change backwards compatible so clients can choose to continue using the old hard-coded origin/port (`http://localhost:8081`). The main reason is that many devs probably pick an example project and start modifying it to fit their needs. If we required `Info.plist` entries for origin/port values those old modified examples would break.

#### Usage in examples

At this point I only modified `UIExplorer` to show that origin/port can be configured. But I could change all examples if needed.

#### Utility function

There are three places where I added code that looks up the values for origin/port from `Info.plist`. That code is fairly trivial but could still be extracted and shared as a utility function. Please advise if that is necessary and let me know where I should add that utility function to.

### CODE REVIEW

**Comment by amasad** [comment]
I believe @frantic is the best person to review this

File: Examples/UIExplorer/UIExplorer.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -572,6 +588,22 @@
...
+			buildActionMask = 2147483647;
+			files = (
+			);
+			inputPaths = (
+			);
+			outputPaths = (
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/bash;
+			shellScript = "export ORIGIN=$(/usr/libexec/PlistBuddy -c \"Print :RCTWebSocketExecutor:origin\" \"${INFOPLIST_FILE}\")\nexport PORT=$(/usr/libexec/PlistBuddy -c \"Print :RCTWebSocketExecutor:port\" \"${INFOPLIST_FILE}\")\n\nif nc -w 5 -z localhost $PORT ; then\n  if ! curl -s \"${ORIGIN}:${PORT}/status\" | grep -q \"packager-status:running\" ; then\n    echo \"Port ${PORT} already in use, packager is either not running or not running correctly\"\n    exit 2\n  fi\nelse\n  # open w/t --args doesn't work:\n  # open $SRCROOT/../../packager/launchPackager.command --args 111 222 || echo \"Can't start packager automatically\"\n  osascript -e 'tell app \"Terminal\"\n    do script \"'$SRCROOT'/../../packager/launchPackager.command --origin='${ORIGIN}' --port='${PORT}'\"\n  end tell' || echo \"Can't start packager automatically\"\nfi";
```
Comment (frantic) [inline Examples/UIExplorer/UIExplorer.xcodeproj/project.pbxproj]: This script is very long, can you extract it to a separate file and call that file from here?

File: Examples/UIExplorer/UIExplorer/AppDelegate.m [hunk 2]
```diff
@@ -36,7 +36,23 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
    * on the same Wi-Fi network.
    */
 
-  jsCodeLocation = [NSURL URLWithString:@"http://localhost:8081/Examples/UIExplorer/UIExplorerApp.includeRequire.runModule.bundle?dev=true"];
+  NSMutableString* url;
+  NSDictionary *dict = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"RCTWebSocketExecutor"];
```
Comment (frantic) [inline Examples/UIExplorer/UIExplorer/AppDelegate.m]: this should definitely be extracted to a function. See #548 + comments

File: packager/packager.js [hunk 3]
```diff
@@ -169,7 +173,7 @@ function getDevToolsLauncher(options) {
       res.writeHead(200, {'Content-Type': 'text/html'});
       fs.createReadStream(debuggerPath).pipe(res);
     } else if (req.url === '/launch-chrome-devtools') {
-      var debuggerURL = 'http://localhost:' + options.port + '/debugger-ui';
+      var debuggerURL = options.origin + ':' + options.port + '/debugger-ui';
```
Comment (frantic) [inline packager/packager.js]: Why do you need this? The URL is served on the same host the server is running. Or am I missing something?

**Comment by frantic** [comment]
Good work, thank you! But I feel like we should probably do the other way around - have React target be pure and ReactWithPackager do what React target does now. @nicklockwood / @tadeuzagallo - do you think my suggestion makes sence?

**Comment by ide** [comment]
@frantic fwiw I agree with that structure, and would take it a step further by skipping ReactWithPackager and just having SampleApp, UIExplorer, TicTacToe, etc... call the packager from their own pbxproj files.

File: packager/packager.js (hunk 3 above)
Comment (amccloud) [inline packager/packager.js]: @frantic if you're on device wouldn't this need to be an ip?

**Comment by bparadie** [comment]
@frantic @ide Thanks for the feedback. This is one of my weekend projects, please expect some delays etc. 
Just to confirm: You and @ide suggest switching the default so `Info.plist` is supported by the `React` target and only the `ReactWithPackager` would launch the packager, correct?

Now, you guys realize that that will break modified old projects. The `react-native` [Get Started](https://facebook.github.io/react-native/docs/getting-started.html#quick-startt) section says:
```
Quick start 
- npm install -g react-native-cli
- react-native init AwesomeProject
```

All those old `AwesomeProjects` will be broken if the `React` target stops launching the packager.  Are you ok with that? Can you just add a 'Breaking Changes' section to the next release and we all move on with our lives?

File: Examples/UIExplorer/UIExplorer.xcodeproj/project.pbxproj (hunk 1 above)
Comment (bparadie) [inline Examples/UIExplorer/UIExplorer.xcodeproj/project.pbxproj]: @frantic Sure, can do. What's a good place to place that shell script? This one?: `./scripts`.

File: Examples/UIExplorer/UIExplorer/AppDelegate.m (hunk 2 above)
Comment (bparadie) [inline Examples/UIExplorer/UIExplorer/AppDelegate.m]: @frantic: Do you have any suggestions for a good place? Where do you guys put utility functions? This one?: `./React/Base/RCTUtils.m`

File: packager/packager.js (hunk 3 above)
Comment (bparadie) [inline packager/packager.js]: @frantic Yes, that's the idea. `options.origin` can be an ip address. This was suggested by @jsierles in https://github.com/facebook/react-native/issues/1429#issuecomment-106905639.

File: packager/packager.js [hunk 4]
```diff
@@ -34,6 +34,10 @@ var parseCommandLine = require('./parseCommandLine.js');
 var webSocketProxy = require('./webSocketProxy.js');
 
 var options = parseCommandLine([{
+  command: 'origin',
+  default: 'http://localhost',
+  type: 'string',
+}, {
```
Comment (ide) [inline packager/packager.js]: fyi "origin" (as in specs like CORS) includes the port number

Comment (bparadie) [inline packager/packager.js]: @ide I wanted to model it after HTML's  `location.origin`. But you are right, `origin` includes the port number. Not sure whether `hostname` is a better name and it doesn't include the `protocol`. Is there a better name than `origin`? I went through the [HTML location properties](http://www.w3schools.com/jsref/obj_location.asp). But nothing seems to fit...

Comment (ide) [inline packager/packager.js]: we could split up protocol and hostname since most of the time (for development) the protocol is http and only the hostname needs to change.

Comment (bparadie) [inline packager/packager.js]: @ide What I am hearing you say is that we should model everything after HTML location and replace `origin` with `hostname` and `protocol`. Is that what you are proposing? If so: I like it!

Comment (ide) [inline packager/packager.js]: Yes! `protocol`, `hostname`, and `port` as three separately configurable fields. Could even use Node's `url.format()` to construct the complete origin string if that helps.

**Comment by bparadie** [comment]
Ok, I've made the following changes:
- replaced 'origin' with 'protocol' + 'hostname' as suggested by @ide 
- moved the debug URL creation code to its own utility function as suggested by @frantic 
- moved launch mechanics out to launchPackager.command as suggested by @frantic 
- removed launch script from React project and added launch scripts to examples as suggested by @ide 

But:
- I had to revert one line from facebook@1acca01, see https://github.com/facebook/react-native/issues/1654
- this change will break old projects.


#### Breaking changes: required/optional changes.

If you have an old project and update your react-native module to this version then your old project will not launch the packager, because I removed the launch script from the React project.

Here are the changes that are required/optional...

##### REQUIRED: Add a launch script
Add a run script to your project with this line 
```bash
$SRCROOT/../../packager/launchPackager.command ${INFOPLIST_FILE}`
```
##### OPTIONAL: Use RCTWebSocketExecutorURL 
Use `RCTWebSocketExecutorURL` from `RCTUtils.h` for creating your `jsCodeLocation` in `AppDelegate.m`, i.e.
```objective-c
jsCodeLocation = RCTWebSocketExecutorURL(@"/Examples/SampleApp/index.ios.bundle");
```
##### OPTIONAL: Add a RCTWebSocketExecutor dictionary 
Add a `RCTWebSocketExecutor` dictionary for custom settings to your `Info.plist`, i.e.
screen shot 2015-06-21 at 1 57 36 pm

Please let me know if this looks good to you. 
If so I'll squash/rebase my PR.
Thx!

**Comment by bparadie** [comment]
Hmm, Travis CI build failed. It looks like I need to add the launch scripts to every test targets. Give me a minute!

**Comment by bparadie** [comment]
One remaining problem is that `packager.json` runs `packager.sh` for the `scripts/start` and `bin/react-native-start` those will always launch the packager with the default settings, that is, port=8081.

**Comment by bparadie** [comment]
@frantic I am a little bit puzzled why Travis CI is failing.

```
2015-06-21 22:27:29.662 [info][tid:com.facebook.React.JavaScript] "Running application "AppEventsTest" with appParams: {"rootTag":1,"initialProps":{}}. __DEV__ === true, development-level warning are ON, performance optimizations are OFF"
2015-06-21 22:27:29.672 [error][tid:com.facebook.React.JavaScript] "Error: Invariant Violation: Application AppEventsTest has not been registered. This is either due to a require() error during initialization or failure to call AppRegistry.registerComponent.
 stack: 
  runApplication          IntegrationTestsA…:36107
  jsCall                  IntegrationTestsA…:7411
  _callFunction           IntegrationTestsA…:7679
  <unknown>               IntegrationTestsA…:7708
  <unknown>               IntegrationTestsA…:7702
  perform                 IntegrationTestsA…:6198
  batchedUpdates          IntegrationTestsA…:14132
  batchedUpdates          IntegrationTestsA…:4741
  <unknown>               IntegrationTestsA…:7701
  applyWithErrorReporter  IntegrationTestsA…:7446
  guardReturn             IntegrationTestsA…:7468
  processBatch            IntegrationTestsA…:7700
 URL: http://localhost:8081/Examples/UIExplorer/UIExplorerIntegrationTests/js/IntegrationTestsApp.includeRequire.runModule.bundle?dev=true
 line: 1762
 message: Invariant Violation: Application AppEventsTest has not been registered. This is either due to a require() error during initialization or failure to call AppRegistry.registerComponent."
```

I even removed my workaround for https://github.com/facebook/react-native/issues/1654 . But that didn't help either. Would you know what's going on here? Thanks!

**Comment by bparadie** [comment]
I can now at least reproduce the problem locally w/t  `rm -rf node_modules && npm install && ./scripts/objc-test.sh`...

**Comment by bparadie** [comment]
Ok, I found the problem (don't ask how I found it): react-native's `package.json` requires version `5.4.3` for the `babel` node nodule. Unfortunately `babel`'s `package.json` sloppily accepts `^5.4.3` for `babel-core`. Now, if you start with a fresh clone (as Travis CI does) you'll end up with `babel 5.4.3` using `babel-core 5.6.2` (which is the latest `babel-core` version). Here is the problem: `babel-core 5.6.2` will break `./scripts/objc-test.sh` tests. The version that works is `babel-core 5.5.6`.

For that reason this PR contains this workaround. In `package.json` I explicitly request `babel-core 5.5.6`:
```javascript
    "babel": "5.4.3",
    "babel-core": "5.5.6",
```

Would somebody please figure out why `babel-core 5.6.2` breaks the build loop?

**Comment by bparadie** [comment]
The build loop failed because the `flow server is busy`:

```
Flow server launched for /private/tmp/react-native-x3yOQW0y/EndToEndTest
Spawned flow server (child pid=2838)
Logs will go to /var/folders/vy/rcv48w3j4w79llzf_x6qnvw40000gn/T/flow/zSprivatezStmpzSreact-native-x3yOQW0yzSEndToEndTest.log
The flow server will be ready in a moment.
Error: flow server is busy
Error: flow server is busy
Error: flow server is busy
No errors!
[Info] Loading settings for scheme 'EndToEndTest' ...
```

That's not really my problem, is it?

**Comment by frantic** [comment]
`flow server is busy` is just a warning, after that it says No errors and continues the build, so that's not the reason for the failure. This is:

```
/Users/travis/Library/Developer/Xcode/DerivedData/EndToEndTest-bmjlkbqwxoavxthdezsighobnyjh/Build/Intermediates/EndToEndTest.build/Debug-iphonesimulator/EndToEndTest.build/Script-4CED847C1B3756CA00715E21.sh: line 2: /tmp/react-native-x3yOQW0y/EndToEndTest/../../packager/launchPackager.command: No such file or directory
```

**Comment by tomstilljay1** [comment]
Speak English or just meet me somewhere to discuss this matter

**Comment by bparadie** [comment]
@frantic Ah, thanks! I'll look into it. BTW, I think none of the PRs pass the build loop right now because of  `babel-core 5.6.2`. Is somebody looking at it?

**Comment by frantic** [comment]
> Here is the problem: babel-core 5.6.2 will break ./scripts/objc-test.sh tests.

I'm looking into this now. What makes you think babel is the reason? Could you please compare packager output bundles?

**Comment by frantic** [comment]
Looks like `displayName` annotations are missing...
image

**Comment by bparadie** [comment]
> What makes you think babel is the reason? 

My local version was not showing the problem. Then I found that the difference was the build loop using `babel-core 5.6.2` while my local version that worked used `babel-core 5.5.6`. When I added `babel-core 5.5.6`  to `package.json` the ./scripts/objc-test.sh passed.

The problem could be something completely different. But that one line makes the problem go away.

**Comment by bparadie** [comment]
Finally, Travis CI passed! Please let me know if you want me to make more changes. Otherwise I'd be ready to rebase/squash. Cheers.

**Comment by bparadie** [comment]
@frantic Ready for review. I've merged the tip of master. Travis CI build passes.

**Comment by bparadie** [comment]
@frantic Is there still interest in this PR? Travis CI passed 17 days ago. I need to know whether there is still interest in my changes before I commit more time into keeping this branch in sync with master.

File: Examples/UIExplorer/UIExplorer/Info.plist [hunk 5]
```diff
@@ -36,13 +34,23 @@
...
 	<key>UIViewControllerBasedStatusBarAppearance</key>
 	<false/>
+	<key>RCTWebSocketExecutor</key>
+	<dict>
+		<key>protocol</key>
+		<string>http:</string>
+		<key>hostname</key>
+		<string>localhost</string>
+		<key>port</key>
+		<integer>8081</integer>
```
Comment (frantic) [inline Examples/UIExplorer/UIExplorer/Info.plist]: Could we just have URL here instead of parts? Seems like in the end these will get glued together anyways. IMHO it's much easier to deal with URLs.

File: React/Base/RCTUtils.h [hunk 6]
```diff
@@ -62,4 +62,7 @@ RCT_EXTERN NSError *RCTErrorWithMessage(NSString *message);
 RCT_EXTERN id RCTNullIfNil(id value);
 RCT_EXTERN id RCTNilIfNull(id value);
 
+// Create NSURL for jsCodeLocation in AppDelegate:application
+RCT_EXTERN NSURL *RCTWebSocketExecutorURL(NSString *appendString);
```
Comment (frantic) [inline React/Base/RCTUtils.h]: Let's rename this to something less WebSocketExecutor-specific. Maybe `RCTPackagerURL`?

small nit: the argument can actually be named "path", because that's what it is.

File: React/Base/RCTUtils.m [hunk 7]
```diff
@@ -286,6 +286,31 @@ id RCTNilIfNull(id value)
   return value == (id)kCFNull ? nil : value;
 }
 
+NSURL *RCTWebSocketExecutorURL(NSString *appendString)
+{
+  NSMutableString* url;
+  NSDictionary *dict = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"RCTWebSocketExecutor"];
+  if( dict )
```
Comment (frantic) [inline React/Base/RCTUtils.m]: nit: see how we format the code in other places. Space after `if`, no spaces within `()`, curly brace on the same line, etc.

File: React/Base/RCTUtils.m [hunk 8]
```diff
@@ -286,6 +286,31 @@ id RCTNilIfNull(id value)
...
+  {
+    NSString* protocol = [dict objectForKey:@"protocol"];
+    NSString* hostname = [dict objectForKey:@"hostname"];
+    NSString* port = [NSString stringWithFormat:@"%@", [dict objectForKey:@"port"]]; ;
+    url = [NSMutableString stringWithString:protocol];
+    [url appendString:@"//"];
+    [url appendString:hostname];
+    [url appendString:@":"];
+    [url appendString:port];
```
Comment (frantic) [inline React/Base/RCTUtils.m]: String concats are very error prone. If you take my suggestion about URL in `Info.plist` you can just do `[NSURL URLWithString]` and then `[url URLByAppendingPathComponent:]`

File: packager/packager.js [hunk 9]
```diff
@@ -35,6 +35,14 @@ var parseCommandLine = require('./parseCommandLine.js');
 var webSocketProxy = require('./webSocketProxy.js');
 
 var options = parseCommandLine([{
+  command: 'protocol',
```
Comment (frantic) [inline packager/packager.js]: what does it actually give us? if we wanted to use `https` we would have to set up certificates. Are there any other cases for this value to not be `http:`?

**Comment by frantic** [comment]
@nicklockwood wondering what do you think about this

File: packager/packager.js (hunk 9 above)
Comment (ide) [inline packager/packager.js]: The protocol will almost always be http (though https enables HTTP/2) but separating it from the hostname makes it easier to configure only the hostname and leave the protocol/port alone, which is a common thing to do.

File: React/Base/RCTUtils.m [hunk 10]
```diff
@@ -286,6 +286,31 @@ id RCTNilIfNull(id value)
   return value == (id)kCFNull ? nil : value;
 }
 
+NSURL *RCTWebSocketExecutorURL(NSString *appendString)
+{
+  NSMutableString* url;
+  NSDictionary *dict = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"RCTWebSocketExecutor"];
+  if( dict )
+  {
+    NSString* protocol = [dict objectForKey:@"protocol"];
```
Comment (ide) [inline React/Base/RCTUtils.m]: nit: `dict[@"protocol"]`

File: React/Base/RCTUtils.m (hunk 7 above)
Comment (ide) [inline React/Base/RCTUtils.m]: also `Type *var`

**Comment by ms88privat** [comment]
Something new here? Today i had a really hard time to figure out how to debug on a real device. In the end i manually changed "localhost" to my ip address in all 4 files. Maybe we should write it to the troubleshooting page.

**Comment by bparadie** [comment]
@ms88privat This PR has been on my TODO list for quite a while. But lately I've been just too busy. There is also one open question whether `protocol`, `hostname`, and `port` should be separate fields.

@ide @frantic: Hey guys, can I get a quick yay/nay on the discussion in https://github.com/facebook/react-native/pull/1546#discussion-diff-34632065 and https://github.com/facebook/react-native/pull/1546#discussion-diff-34632468 ?
In a nutshell @frantic questions whether it makes sense to have `protocol`, `hostname`, and `port` as separate fields. Instead he suggests we should just use one URL. I think it will be better to have separate fields. 
Final vote?

**Comment by ide** [comment]
Either approach is fine with me since neither really blocks the other. I do like being able to configure just the hostname because I use mycomputername.local instead of localhost for example.

**Comment by bparadie** [comment]
@frantic Hi, I have made all the additional changes as requested:
- Info.plist: `url` instead of `protocol` plus `hostname`
- renamed `RCTWebSocketExecutorURL` to `RCTPackagerURL`
I fixed the nits too.
All tests pass. 
Are we getting closer?

File: React/Views/RCTViewManager.m [hunk 11]
```diff
@@ -54,7 +54,7 @@ - (dispatch_queue_t)methodQueue
   return _bridge.uiManager.methodQueue;
 }
 
-- (UIView *)viewWithProps:(NSDictionary *)props
+- (UIView *)viewWithProps:(NSDictionary *) __attribute__((unused)) props
```
Comment (bparadie) [inline React/Views/RCTViewManager.m]: That `unused variable` warning was getting on my nerves...

File: local-cli/generator-ios/templates/xcodeproj/project.pbxproj [hunk 12]
```diff
@@ -513,6 +513,35 @@
...
+			);
+			inputPaths = (
+			);
+			outputPaths = (
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/bash;
+			shellScript = "if [ -d \"$SRCROOT/../../packager\" ]\nthen\n    PACKAGER_ROOT=$SRCROOT/../../packager\nelse\n    PACKAGER_ROOT=$SRCROOT/node_modules/react-native/packager\nfi\n\n$PACKAGER_ROOT/launchPackager.command ${INFOPLIST_FILE}\n";
+		};
+/* End PBXShellScriptBuildPhase section */
```
Comment (bparadie) [inline local-cli/generator-ios/templates/xcodeproj/project.pbxproj]: I am assuming this is the right thing to do. If this template is being used to create new iOS projects then those projects need the same script.

File: React/Views/RCTViewManager.m (hunk 11 above)
Comment (nicklockwood) [inline React/Views/RCTViewManager.m]: We've fixed this already internally, but for future reference you can just say:

  - (UIView *)viewWithProps:(__unused NSDictionary *)props

Comment (bparadie) [inline React/Views/RCTViewManager.m]: @nicklockwood Oh yeah, that's much better!

**Comment by jsierles** [comment]
This looks good! Would love to get this merged so we can start building some tooling around it.

**Comment by bparadie** [comment]
@jsierles Me too! It is unclear to me who is assigned to this PR. Is it @frantic ?
(And in the meantime this branch got out of sync again...)

**Comment by bparadie** [comment]
It looks like the URL schema for loading the JS files has changed. But I think somebody missed to adjust the `Movies` and `SampleApp` examples... What about the iOS templates?

**Comment by bparadie** [comment]
Tests are passing But I didn't do anything except for merging master:
```
Merge made by the 'recursive' strategy.
 website/jsdocs/jsdocs.js             |  29 ++++++++----
 website/layout/AutodocsLayout.js     | 142 ++++++++++++++++++++++++++++++++++++++++----------------
 website/server/extractDocs.js        |   3 +-
 website/src/react-native/showcase.js |   6 +++
 4 files changed, 130 insertions(+), 50 deletions(-)
```

I don't see how those changes could fix the build loop.
Is it possible that the build/test loop sometimes just randomly fails?

**Comment by bparadie** [comment]
@frantic @ide @nicklockwood Would anybody of you be so kind and review this PR? Thanks!

**Comment by bparadie** [comment]
@frantic @ide @nicklockwood It looks like nobody is interested in this PR. I'll leave it open for another 7 days. But then I'll close this PR if I don't hear back from you guys.

**Comment by brentvatne** [comment]
@bparadie - can you rebase?

@nicklockwood - do you have any time for this?

cc @vjeux this looks useful

**Comment by jsierles** [comment]
Just to add my 2c here. This issue is important because it stops developers from getting on board as quickly as possible with debugging from devices. We still have to manually update RCTWebSocketManager.m, and these changes are overwritten when updating React Native. Now that hot reloading is a reality in React Native projects, it's even more important.

**Comment by bparadie** [comment]
Time is up. I am closing this PR.

**Comment by jsierles** [comment]
Hmm, too bad on this one. Given android is out now, maybe it's time to look at a single cross platform solution that doesn't depend on plist files or anything platform-specific. A global config file - maybe package.json?

**Comment by PaulBGD** [comment]
Sad, this is something that would be really useful.

**Comment by qimingweng** [comment]
I've posted this on product pains so it would be great if you could upvote it there and show the react-native team that this is an important feature https://productpains.com/post/react-native/allow-packager-port-to-be-configurable-change-from-8081/
